File: 轮廓检测.py
import cv2 as cv
import numpy as np
from tqdm import tqdm
from scipy.signal import find_peaks


def calculate_threshold(red_channel, mask=None):
    """
    计算阈值，阈值为出现频率最高的两个峰值之间的最低值。

    参数:
    red_channel (numpy.ndarray): 输入的红色通道图像。
    mask (numpy.ndarray, optional): 可选的掩码图像，用于指定计算阈值的区域，形状必须与 red_channel 相同。

    返回:
    int: 计算得到的阈值。

    异常:
    ValueError: 如果无法找到两个峰值，则抛出此异常。
    """
    if mask is not None and mask.shape == red_channel.shape:
        red_channel = cv.bitwise_and(red_channel, red_channel, mask=mask)
    hist = cv.calcHist([red_channel], [0], None, [256], [0, 256]).flatten()
    hist[0] = 0
    peaks, _ = find_peaks(hist)
    sorted_peaks = sorted(peaks, key=lambda x: hist[x], reverse=True)

    # 找到出现频率最高的两个峰值
    if len(sorted_peaks) < 2:
        raise ValueError("无法找到两个峰值")

    second_peak_index = min(sorted_peaks[:2])
    min_val_after_peak = np.argmin(hist[second_peak_index:]) + second_peak_index
    return min_val_after_peak


def segment_image(src, threshold, kernel_size):
    """
    对图像进行分割，并返回分割后的边界框和掩码。

    参数:
    src (numpy.ndarray): 输入的源图像。
    threshold (int): 用于二值化的阈值。
    kernel_size (tuple): 用于形态学操作的内核大小。

    返回:
    tuple: 包含两个元素的元组：
        - (x, y, w, h) (tuple): 分割后的边界框，包含左上角坐标 (x, y) 和宽度、高度 (w, h)。
        - mask (numpy.ndarray or None): 分割后的掩码图像，如果没有找到轮廓则为 None。

    异常:
    无
    """
    _, src_bin = cv.threshold(src, threshold, 255, cv.THRESH_BINARY)
    kernel = cv.getStructuringElement(cv.MORPH_RECT, kernel_size)
    src_bin = cv.morphologyEx(src_bin, cv.MORPH_OPEN, kernel)
    src_bin = cv.morphologyEx(src_bin, cv.MORPH_CLOSE, kernel)
    coords = cv.findNonZero(src_bin)
    contours, _ = cv.findContours(src_bin, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
    if not contours:
        return cv.boundingRect(coords), None
    contour = max(contours, key=cv.contourArea)
    mask = np.zeros(src.shape, dtype=np.uint8)
    cv.drawContours(mask, [contour], -1, 1, thickness=cv.FILLED)
    return cv.boundingRect(coords), mask


def crop_to_bbox(frame, bbox):
    """
    根据给定的边界框裁剪图像。

    参数:
    frame (numpy.ndarray): 输入的图像。
    bbox (tuple): 边界框，包含左上角坐标 (x, y) 和宽度、高度 (w, h)。

    返回:
    numpy.ndarray: 裁剪后的图像（深拷贝）。
    """
    x, y, w, h = bbox
    return frame[y : y + h, x : x + w].copy()


def classify_contour(contour, binary_image, thickness=20):
    """
    根据轮廓内像素的平均值对轮廓进行分类。

    参数:
    contour (numpy.ndarray): 输入的轮廓。
    binary_image (numpy.ndarray): 输入的二值化图像。
    thickness (int, optional): 用于绘制轮廓的厚度。默认值为 20。

    返回:
    tuple: 轮廓的颜色，根据平均值分类为绿色 (0, 255, 0) 或蓝色 (255, 0, 0)。
    """
    # 创建与 binary_image 大小相同的掩码
    mask = np.zeros_like(binary_image)
    cv.drawContours(mask, [contour], -1, 255, thickness)

    # 创建轮廓内涂满的图像
    filled_contour_image = np.zeros_like(binary_image)
    cv.drawContours(filled_contour_image, [contour], -1, 255, thickness=cv.FILLED)

    # 获取在掩码图像和轮廓内涂满的图像中都为255的点
    combined_mask = cv.bitwise_and(mask, filled_contour_image)

    # 获取这些点在原来的二值化图像中的颜色
    masked_points = binary_image[combined_mask == 255]

    # 计算这些点的像素平均值
    avg_color = np.mean(masked_points)

    # 根据平均值对边缘分类
    if avg_color < 128:
        color = (0, 255, 0)  # 红色圆环外侧边缘为绿色
    else:
        color = (255, 0, 0)  # 红色圆环内侧边缘为蓝色
    return color


def process_frame(
    frame,
    previous_bbox_area=0,
    kernel_size=(9, 9),
    center_color_threshold=0,
    min_length=500,
    circularity_threshold=0.7,
    area_threshold=1000,
    region_size=10,
    center_circle_diameter=50,
    center_circle_thickness=2,
):
    """
    处理单帧图像，检测轮廓并计算公共中心。

    必填参数:
    frame (numpy.ndarray): 输入的图像帧。

    可选参数:
    计算相关:
    previous_bbox_area (float): 前一帧的边界框面积。默认值为 0。
    kernel_size (tuple): 用于形态学操作的内核大小。默认值为 (9, 9)。
    center_color_threshold (int): 判断中心颜色的阈值。默认值为 0。
    min_length (int): 轮廓的最小长度阈值。默认值为 500。
    circularity_threshold (float): 轮廓的圆形度阈值。默认值为 0.7。
    area_threshold (int): 视为同一轮廓的最大面积差距阈值。默认值为 1000。
    region_size (int): 判断中心颜色的判定区域大小。默认值为 10。

    绘图相关:
    center_circle_diameter (int): 画出公共中心时的直径。默认值为 50。
    center_circle_thickness (int): 画出公共中心时的线宽。默认值为 2。

    返回:
    tuple: 包含四个元素的元组：
        - processed_frame (numpy.ndarray): 处理后的图像帧。
        - current_bbox_area (float): 当前帧的边界框面积。
        - contours_info (list): 包含轮廓面积和颜色信息的列表。
        - center_color (int or None): 公共中心的颜色，红色为 255，黑色为 0，无效为 None。
    """
    red_channel = frame[:, :, 2]
    threshold = calculate_threshold(red_channel)

    bbox, mask = segment_image(red_channel, threshold, (5, 5))
    current_bbox_area = bbox[2] * bbox[3]

    if previous_bbox_area > 0 and current_bbox_area > previous_bbox_area * 1.5:
        bbox, mask = segment_image(red_channel, threshold, (7, 7))
        current_bbox_area = bbox[2] * bbox[3]

    if current_bbox_area == 0:
        return frame, previous_bbox_area, [], None

    red_channel = crop_to_bbox(red_channel, bbox)
    mask = crop_to_bbox(mask, bbox)
    cropped_frame = crop_to_bbox(frame, bbox)

    threshold = calculate_threshold(red_channel, mask)
    _, binary_image = cv.threshold(red_channel, threshold, 255, cv.THRESH_BINARY_INV)

    kernel = cv.getStructuringElement(cv.MORPH_RECT, kernel_size)
    processed_image = cv.morphologyEx(binary_image, cv.MORPH_OPEN, kernel)
    processed_image = cv.morphologyEx(processed_image, cv.MORPH_CLOSE, kernel)

    # 轮廓发现
    edges = cv.Canny(processed_image, 50, 150)
    contours, _ = cv.findContours(edges, cv.RETR_LIST, cv.CHAIN_APPROX_SIMPLE)

    # 计算每个轮廓的面积和圆形度
    max_radius = min(bbox[2:]) // 2
    max_area = np.pi * (max_radius**2)

    # 按照面积顺序处理轮廓
    contours = sorted(contours, key=cv.contourArea)

    contours_info = []
    filtered_contours = []
    pre_area = 0
    for contour in contours:
        area = cv.contourArea(contour)
        perimeter = cv.arcLength(contour, True)
        if perimeter < min_length or area > max_area:
            continue
        circularity = 4 * np.pi * (area / (perimeter * perimeter))
        if circularity >= circularity_threshold:
            if pre_area != 0 and abs(area - pre_area) < area_threshold:
                continue
            pre_area = area
            color = classify_contour(contour, binary_image)
            contours_info.append((area, color))
            filtered_contours.append(contour)
            cv.drawContours(cropped_frame, [contour], -1, color, 2)

    # 计算公共中心
    if filtered_contours:
        centers = [cv.moments(contour) for contour in filtered_contours]
        cx = int(sum(m["m10"] / m["m00"] for m in centers) / len(centers))
        cy = int(sum(m["m01"] / m["m00"] for m in centers) / len(centers))

        # 判断公共中心邻域范围内的像素点颜色
        region = binary_image[
            max(0, cy - region_size) : min(binary_image.shape[0], cy + region_size),
            max(0, cx - region_size) : min(binary_image.shape[1], cx + region_size),
        ]
        avg_color = np.mean(region)

        if avg_color < 128 - center_color_threshold:
            center_color = 255  # 红色
            border_color = (0, 255, 0)  # 绿色
        elif avg_color > 128 + center_color_threshold:
            center_color = 0  # 黑色
            border_color = (255, 0, 0)  # 蓝色
        else:
            center_color = None  # 无效
            border_color = (255, 255, 255)  # 白色

        # 画出公共中心
        cv.circle(
            cropped_frame,
            (cx, cy),
            center_circle_diameter // 2,
            border_color,
            center_circle_thickness,
        )

        return cropped_frame, current_bbox_area, contours_info, center_color
    return cropped_frame, current_bbox_area, contours_info, None


class ContourDetection:
    DEFAULT_SETTINGS = {
        "kernel_size": (9, 9),
        "center_color_threshold": 0,
        "min_length": 500,
        "circularity_threshold": 0.7,
        "area_threshold": 1000,
        "region_size": 10,
        "center_circle_diameter": 50,
        "center_circle_thickness": 2,
        "fps": 10,
        "mode": "zoom_in",
    }

    def __init__(self, settings=None):
        """
        初始化 ContourDetection 类。

        参数:
        settings (dict, optional): 包含设置参数的字典。默认值为 None。
        """
        self.settings = self.DEFAULT_SETTINGS.copy()
        if settings:
            self.settings.update(settings)
        self.reset_counters()
        self.reset_state()

    def reset_counters(self):
        """
        重置计数器。
        """
        self.green_count = 0
        self.blue_count = 0
        self.white_count = 0
        self.count_value = 0
        self.total_frames = 0

    def reset_state(self):
        """
        重置状态属性。
        """
        self.previous_bbox_area = 0
        self.previous_contours_info = []
        self.previous_center_color = None
        self.status = None

    def update(self, frame):
        """
        更新并处理当前帧。

        参数:
        frame (numpy.ndarray): 当前最新的帧。
        """
        self.total_frames += 1
        (
            processed_frame,
            self.previous_bbox_area,
            current_contours_info,
            current_center_color,
        ) = process_frame(
            frame,
            self.previous_bbox_area,
            self.settings["kernel_size"],
            self.settings["center_color_threshold"],
            self.settings["min_length"],
            self.settings["circularity_threshold"],
            self.settings["area_threshold"],
            self.settings["region_size"],
            self.settings["center_circle_diameter"],
            self.settings["center_circle_thickness"],
        )

        # 匹配轮廓并判断放大或缩小
        scale_count = 0
        area_threshold = 50000  # 面积变化阈值
        if self.previous_contours_info and current_contours_info:
            for prev_area, prev_color in self.previous_contours_info:
                for curr_area, curr_color in current_contours_info:
                    if (
                        prev_color == curr_color
                        and abs(curr_area - prev_area) < area_threshold
                    ):
                        if curr_area > prev_area:
                            scale_count += 1
                        else:
                            scale_count -= 1
        if scale_count > 0:
            border_color = (0, 255, 0)  # 绿色
            self.green_count += 1
            self.status = "zoom_out"
        elif scale_count < 0:
            border_color = (255, 0, 0)  # 蓝色
            self.blue_count += 1
            self.status = "zoom_in"
        else:
            border_color = (255, 255, 255)  # 默认白色
            self.white_count += 1

        border_thickness = min(10 + abs(scale_count), 50)

        # 判断计数模式和更新计数值
        if (
            self.previous_center_color is not None
            and current_center_color is not None
            and self.status is not None
        ):
            if (
                self.previous_center_color == 255
                and current_center_color == 0
                and self.settings["mode"] == self.status
            ):
                self.count_value += 1
            if (
                self.previous_center_color == 0
                and current_center_color == 255
                and self.settings["mode"] != self.status
            ):
                self.count_value -= 1

        # 给画面增加边框
        processed_frame = cv.copyMakeBorder(
            processed_frame,
            border_thickness,
            border_thickness,
            border_thickness,
            border_thickness,
            cv.BORDER_CONSTANT,
            value=border_color,
        )

        frame[: processed_frame.shape[0], : processed_frame.shape[1]] = processed_frame

        # 在整体的右上角打印计数值
        cv.putText(
            frame,
            str(self.count_value),
            (frame.shape[1] - 400, 300),  # 右上角位置
            cv.FONT_HERSHEY_SIMPLEX,
            5,
            (255, 255, 255),
            10,
        )

        self.previous_contours_info = current_contours_info
        self.previous_center_color = current_center_color

        return frame

    def process_video(self, input_path, output_path):
        """
        处理视频文件，用于测试效果。

        参数:
        input_path (str): 输入视频文件路径。
        output_path (str): 输出视频文件路径。
        """
        cap = cv.VideoCapture(input_path)
        # ret, frame = cap.read()
        # original_height, original_width = frame.shape[:2]
        # fourcc = cv.VideoWriter_fourcc(*"mp4v")
        # out = cv.VideoWriter(
        # output_path, fourcc, self.settings["fps"], (original_width, original_height)
        # )
        frame_count = int(cap.get(cv.CAP_PROP_FRAME_COUNT))

        self.reset_counters()
        self.reset_state()

        for i in tqdm(range(frame_count), desc="Processing video"):
            ret, frame = cap.read()
            if not ret:
                break
            frame = self.update(frame)
            # out.write(frame)

        cap.release()
        # out.release()

        # 打印统计信息
        total_frames = self.green_count + self.blue_count + self.white_count
        print(
            f"绿色边框帧数: {self.green_count} ({self.green_count / total_frames * 100:.2f}%)"
        )
        print(
            f"蓝色边框帧数: {self.blue_count} ({self.blue_count / total_frames * 100:.2f}%)"
        )
        print(
            f"白色边框帧数: {self.white_count} ({self.white_count / total_frames * 100:.2f}%)"
        )
        print(f"总圈数: {self.count_value}")


def main():
    video_path = "/home/july/physic/test/真实场景.mp4"
    output_path = "轮廓检测+计数（python版）.mp4"

    # 创建 ContourDetection 实例
    contour_detection = ContourDetection()

    # 处理视频文件
    contour_detection.process_video(video_path, output_path)


if __name__ == "__main__":
    main()

File: 轮廓检测.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <string>
#include <numeric>
#include <iostream>

// 计数信息结构体
struct CountInfo {
    int green_count;
    int blue_count;
    int white_count;
    int count_value;
    int total_frames;
};

// 状态信息结构体
struct StateInfo {
    float bbox_area;                                   // 有效区域面积
    std::vector<std::pair<float, bool>> contours_info; // 轮廓面积以及是否为圆环外侧边缘
    int center_color;                                  // 圆心颜色, 1 为红色, 0 为黑色
    int status;                                        // 当前帧的状态, 0 为 "收缩", 1 为 "扩张"
    cv::Point center;                                  // 圆心在有效区域中的坐标
};

// 默认设置
const std::map<std::string, int> DEFAULT_SETTINGS = {
    {"kernel_size", 9},            // 形态学分析去除噪点的卷积核大小
    {"center_color_threshold", 0}, // 中心颜色视为未知的阈值
    {"min_length", 500},           // 有效轮廓的最小长度
    {"circularity_threshold", 70}, // 有效轮廓的圆度阈值（百分比）
    {"area_threshold", 1000},      // 有效轮廓的面积阈值
    {"region_size", 10},           // 中心颜色检测区域大小
    {"mode", 0}                    // 0 为 "收缩计数模式", 1 为 "扩张计数模式"
};

class ContourDetection {
public:
    ContourDetection(const std::map<std::string, int> &settings = DEFAULT_SETTINGS);
    void reset_counters();
    void reset_state();
    cv::Mat process_frame(const cv::Mat &frame);
    void print_result();
    int get_count_value() const { return count_info.count_value; }

private:
    std::map<std::string, int> settings;
    CountInfo count_info;
    StateInfo state_info;

    int calculate_threshold(const cv::Mat &red_channel, const cv::Mat &mask = cv::Mat());
    std::pair<cv::Rect, cv::Mat> segment_image(const cv::Mat &src, int threshold, const cv::Size &kernel_size);
    bool classify_contour(const std::vector<cv::Point> &contour, const cv::Mat &binary_image, int thickness = 20);
};

ContourDetection::ContourDetection(const std::map<std::string, int> &settings) {
    this->settings = DEFAULT_SETTINGS;
    for (const auto &setting : settings) {
        this->settings[setting.first] = setting.second;
    }
    reset_counters();
    reset_state();
}

void ContourDetection::reset_counters() {
    count_info.green_count = 0;
    count_info.blue_count = 0;
    count_info.white_count = 0;
    count_info.count_value = 0;
    count_info.total_frames = 0;
}

void ContourDetection::reset_state() {
    state_info.bbox_area = 0;
    state_info.contours_info.clear();
    state_info.center_color = -1;
    state_info.status = -1;
    state_info.center = cv::Point(-1, -1);
}

// 计算阈值函数
int ContourDetection::calculate_threshold(const cv::Mat &red_channel, const cv::Mat &mask) {
    cv::Mat masked_red_channel;
    if (!mask.empty() && mask.size() == red_channel.size()) {
        cv::bitwise_and(red_channel, red_channel, masked_red_channel, mask);
    } else {
        masked_red_channel = red_channel;
    }

    // 计算直方图
    int histSize = 256;
    float range[] = {0, 256};
    const float *histRange = {range};
    cv::Mat hist;
    cv::calcHist(&masked_red_channel, 1, 0, cv::Mat(), hist, 1, &histSize, &histRange, true, false);
    hist.at<float>(0) = 0; // 忽略第一个bin

    // 找到峰值
    std::vector<int> peaks;
    for (int i = 1; i < histSize - 1; ++i) {
        if (hist.at<float>(i) > hist.at<float>(i - 1) && hist.at<float>(i) > hist.at<float>(i + 1)) {
            peaks.push_back(i);
        }
    }

    if (peaks.size() < 2) {
        throw std::runtime_error("无法找到两个峰值");
    }

    std::sort(peaks.begin(), peaks.end(), [&hist](int a, int b) {
        return hist.at<float>(a) > hist.at<float>(b);
    });

    int second_peak_index = std::min(peaks[0], peaks[1]);
    int min_val_after_peak = std::distance(hist.begin<float>(), std::min_element(hist.begin<float>() + second_peak_index, hist.end<float>()));
    return min_val_after_peak;
}

// 图像分割函数
std::pair<cv::Rect, cv::Mat> ContourDetection::segment_image(const cv::Mat &src, int threshold, const cv::Size &kernel_size) {
    cv::Mat src_bin;
    cv::threshold(src, src_bin, threshold, 255, cv::THRESH_BINARY);
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, kernel_size);
    cv::morphologyEx(src_bin, src_bin, cv::MORPH_OPEN, kernel);
    cv::morphologyEx(src_bin, src_bin, cv::MORPH_CLOSE, kernel);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(src_bin, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    if (contours.empty()) {
        cv::Mat coords;
        cv::findNonZero(src_bin, coords);
        return {cv::boundingRect(coords), cv::Mat()};
    }

    auto contour = *std::max_element(contours.begin(), contours.end(), [](const std::vector<cv::Point> &a, const std::vector<cv::Point> &b) {
        return cv::contourArea(a) < cv::contourArea(b);
    });

    cv::Mat mask = cv::Mat::zeros(src.size(), CV_8U);
    cv::drawContours(mask, std::vector<std::vector<cv::Point>>{contour}, -1, 1, cv::FILLED);
    return {cv::boundingRect(contour), mask};
}

// 轮廓分类函数
bool ContourDetection::classify_contour(const std::vector<cv::Point> &contour, const cv::Mat &binary_image, int thickness) {
    cv::Mat mask = cv::Mat::zeros(binary_image.size(), CV_8U);
    cv::drawContours(mask, std::vector<std::vector<cv::Point>>{contour}, -1, 255, thickness);

    cv::Mat filled_contour_image = cv::Mat::zeros(binary_image.size(), CV_8U);
    cv::drawContours(filled_contour_image, std::vector<std::vector<cv::Point>>{contour}, -1, 255, cv::FILLED);

    cv::Mat combined_mask;
    cv::bitwise_and(mask, filled_contour_image, combined_mask);

    cv::Scalar avg_color = cv::mean(binary_image, combined_mask);

    if (avg_color[0] < 128) {
        return true; // 红色圆环外侧边缘
    } else {
        return false; // 红色圆环内侧边缘
    }
}

cv::Mat ContourDetection::process_frame(const cv::Mat &frame) {
    // 提取红色通道
    cv::Mat red_channel = frame.clone();
    cv::extractChannel(frame, red_channel, 2);

    // 提取有效区域
    int threshold = calculate_threshold(red_channel);
    auto [bbox, mask] = segment_image(red_channel, threshold, cv::Size(5, 5));
    float bbox_area = bbox.area();

    // 动态控制形态学操作的核大小
    if (state_info.bbox_area > 0 && bbox_area > state_info.bbox_area * 1.5) {
        std::tie(bbox, mask) = segment_image(red_channel, threshold, cv::Size(7, 7));
        bbox_area = bbox.area();
    }

    // 如果没有有效区域，则当前帧无效(罕见情况！)
    if (bbox_area == 0) {
        return frame;
    }

    // 开始处理当前帧
    count_info.total_frames++;
    state_info.bbox_area = bbox_area;

    // 裁剪图像
    red_channel = red_channel(bbox);
    mask = mask(bbox);
    cv::Mat cropped_frame = frame(bbox);

    // 生成逆二值化图像，此时0代表红色，255代表黑色背景
    threshold = calculate_threshold(red_channel, mask);
    cv::Mat binary_image;
    cv::threshold(red_channel, binary_image, threshold, 255, cv::THRESH_BINARY_INV);

    // 通过形态学操作去除噪点
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(settings["kernel_size"], settings["kernel_size"]));
    cv::morphologyEx(binary_image, binary_image, cv::MORPH_OPEN, kernel);
    cv::morphologyEx(binary_image, binary_image, cv::MORPH_CLOSE, kernel);

    // 轮廓发现
    cv::Mat edges;
    cv::Canny(binary_image, edges, 50, 150);
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(edges, contours, cv::RETR_LIST, cv::CHAIN_APPROX_SIMPLE);

    // 计算理论最大轮廓面积
    int max_radius = std::min(bbox.width, bbox.height) / 2;
    float max_area = CV_PI * (max_radius * max_radius);

    // 按照面积排序
    std::sort(contours.begin(), contours.end(), [](const std::vector<cv::Point> &a, const std::vector<cv::Point> &b) {
        return cv::contourArea(a) < cv::contourArea(b);
    });

    // 提取有效轮廓信息
    std::vector<std::pair<float, bool>> contours_info;
    std::vector<cv::Moments> centers;
    float pre_area = 0;
    for (const auto &contour : contours) {
        float area = cv::contourArea(contour);
        float perimeter = cv::arcLength(contour, true);
        if (perimeter < settings["min_length"] || area > max_area) {
            continue;
        }
        float circularity = 4 * CV_PI * (area / (perimeter * perimeter));
        if (circularity >= settings["circularity_threshold"] / 100.0) {
            if (pre_area != 0 && std::abs(area - pre_area) < settings["area_threshold"]) {
                continue;
            }
            pre_area = area;
            bool isOuterEdge = classify_contour(contour, binary_image);
            contours_info.emplace_back(area, isOuterEdge);
            centers.push_back(cv::moments(contour));
            cv::Scalar color = isOuterEdge ? cv::Scalar(0, 255, 0) : cv::Scalar(255, 0, 0);
            cv::drawContours(cropped_frame, std::vector<std::vector<cv::Point>>{contour}, -1, color, 2);
        }
    }

    // 未发现轮廓(常见情况)
    if (contours_info.empty()) {
        state_info.contours_info.clear();
        state_info.center_color = -1;
        state_info.status = -1;
        count_info.white_count++;
        return cropped_frame;
    }

    // 计算公共中心
    int cx = static_cast<int>(std::accumulate(centers.begin(), centers.end(), 0.0, [](double sum, const cv::Moments &m) {
                                  return sum + m.m10 / m.m00;
                              }) /
                              centers.size());
    int cy = static_cast<int>(std::accumulate(centers.begin(), centers.end(), 0.0, [](double sum, const cv::Moments &m) {
                                  return sum + m.m01 / m.m00;
                              }) /
                              centers.size());

    // 判断公共中心邻域范围内的像素点颜色
    cv::Rect region_rect(cx - settings["region_size"], cy - settings["region_size"], settings["region_size"] * 2, settings["region_size"] * 2);
    cv::Mat region = binary_image(region_rect);
    double avg_color = cv::mean(region)[0];

    state_info.center = cv::Point(cx, cy);

    int center_color;
    if (avg_color < 128 - settings["center_color_threshold"]) {
        center_color = 1; // 红色
    } else if (avg_color > 128 + settings["center_color_threshold"]) {
        center_color = 0; // 黑色
    } else {
        center_color = -1; // 无效
    }

    // 匹配轮廓并判断放大或缩小
    int scale_count = 0;
    int area_threshold = 50000; // 面积变化阈值
    if (!state_info.contours_info.empty() && !contours_info.empty()) {
        for (const auto &[prev_area, prev_type] : state_info.contours_info) {
            for (const auto &[curr_area, curr_type] : contours_info) {
                if (prev_type == curr_type && std::abs(curr_area - prev_area) < area_threshold) {
                    if (curr_area > prev_area) {
                        scale_count++;
                    } else {
                        scale_count--;
                    }
                }
            }
        }
    }

    cv::Scalar border_color;
    if (scale_count > 0) {
        border_color = cv::Scalar(0, 255, 0); // 绿色
        count_info.green_count++;
        state_info.status = 1; // zoom_out
    } else if (scale_count < 0) {
        border_color = cv::Scalar(255, 0, 0); // 蓝色
        count_info.blue_count++;
        state_info.status = 0; // zoom_in
    } else {
        border_color = cv::Scalar(255, 255, 255); // 默认白色
        count_info.white_count++;
        // state_info.status = -1; // 无效状态
    }
    cv::copyMakeBorder(cropped_frame, cropped_frame, 10, 10, 10, 10, cv::BORDER_CONSTANT, border_color);

    // 判断计数模式和更新计数值
    if (state_info.center_color != -1 && center_color != -1 && state_info.status != -1) {
        if (state_info.center_color == 1 && center_color == 0 && settings["mode"] == state_info.status) {
            count_info.count_value++;
        }
        if (state_info.center_color == 0 && center_color == 1 && settings["mode"] != state_info.status) {
            count_info.count_value--;
        }
    }

    state_info.contours_info = contours_info;
    state_info.center_color = center_color;
    return cropped_frame;
}

void ContourDetection::print_result() {
    int total_frames = count_info.green_count + count_info.blue_count + count_info.white_count;
    if (total_frames == 0) {
        std::cout << "没有处理任何帧。" << std::endl;
        return;
    }

    std::cout << "绿色边框帧数: " << count_info.green_count << " (" << std::fixed << std::setprecision(2) << (count_info.green_count / static_cast<float>(total_frames) * 100) << "%)" << std::endl;
    std::cout << "蓝色边框帧数: " << count_info.blue_count << " (" << std::fixed << std::setprecision(2) << (count_info.blue_count / static_cast<float>(total_frames) * 100) << "%)" << std::endl;
    std::cout << "白色边框帧数: " << count_info.white_count << " (" << std::fixed << std::setprecision(2) << (count_info.white_count / static_cast<float>(total_frames) * 100) << "%)" << std::endl;
    std::cout << "总圈数: " << count_info.count_value << std::endl;
}

void print_progress(float progress, int current_frame, int total_frames, double elapsed_time, double estimated_time) {
    int bar_width = 50;
    std::cout << "\rProcessing video: ";
    std::cout << std::fixed << std::setprecision(2) << progress << "% [";
    int pos = bar_width * progress / 100.0;
    for (int i = 0; i < bar_width; ++i) {
        if (i < pos)
            std::cout << "#";
        else
            std::cout << "-";
    }
    std::cout << "] " << current_frame << "/" << total_frames;
    std::cout << " [" << std::fixed << std::setprecision(2) << elapsed_time << "s<" << estimated_time << "s, " << (elapsed_time / current_frame) << "s/it]";
    std::cout.flush();
}

int main() {
    std::string video_path = "/home/july/physic/test/真实场景.mp4";

    cv::VideoCapture cap(video_path);

    cv::Mat frame;
    int frame_count = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_COUNT));

    ContourDetection contour_detection = ContourDetection();

    auto start_time = std::chrono::steady_clock::now();

    std::vector<std::pair<int, double>> count_timestamps; // 记录计数值和时间戳

    for (int i = 0; i < frame_count; ++i) {
        cap >> frame;
        if (frame.empty()) {
            break;
        }
        contour_detection.process_frame(frame);

        // 更新进度条
        float progress = (i + 1) / static_cast<float>(frame_count) * 100;
        auto current_time = std::chrono::steady_clock::now();
        double elapsed_time = std::chrono::duration_cast<std::chrono::seconds>(current_time - start_time).count();
        double estimated_time = (elapsed_time / (i + 1)) * (frame_count - (i + 1));
        print_progress(progress, i + 1, frame_count, elapsed_time, estimated_time);

        // 记录当前计数值和时间戳
        count_timestamps.emplace_back(contour_detection.get_count_value(), elapsed_time);
    }

    std::cout << std::endl;

    contour_detection.print_result();

    // 计算最大速率
    double max_rate = 0.0;
    for (size_t i = 1; i < count_timestamps.size(); ++i) {
        int count_diff = count_timestamps[i].first - count_timestamps[i - 1].first;
        double time_diff = count_timestamps[i].second - count_timestamps[i - 1].second;
        if (time_diff > 0) {
            double rate = count_diff / time_diff;
            if (rate > max_rate) {
                max_rate = rate;
            }
        }
    }

    std::cout << "最大吞吐计数速率: " << max_rate << " 次/秒" << std::endl;

    cap.release();
}